### PR TITLE
Feat/websocket

### PR DESCRIPTION
**Pull request by maxmanukian**
- Added ESModules
  - Implemented a WebSocket server to allow users to connect and receive real-time messages. 
  - Added `ZmqConnection` to connect to and listen for blockchain events.
  - Added functionality using the Dash Core library to decode received hex transactions for easier processing.
  - Implemented a function to modify and improve the reusability of decoded transaction objects.

### CODE REVIEW

File: src/helper.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import crypto from 'crypto'
+import bs58check from 'bs58check'
+
+export const decodeRawTransaction = async (rawtx) =>{
```
Comment (pshenmic) [inline src/helper.js]: I think it would be better to make a separate layer for an API calls. Let's say an DashCoreAPI, which contains all necessary methods to the Core

File: src/messageManager.js
Comment (pshenmic) [inline src/messageManager.js:1]: MassageManager 🙈 =D

File: src/server.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import {WebSocketServer} from 'ws';
+import { v4 as uuid } from 'uuid';
+import ZmqConnection from "./zmq.js";
+import {decodeRawTransaction} from "./helper.js";
+import MessageManager from "./messageManager.js";
+
+export default class Server {
+    constructor() {
+        this.connections = []
```
Comment (pshenmic) [inline src/server.js]: Define them in the class, also we can mark fields private (if they are not used outside this module)

```
export default class Server {
  #connections = []
  #transactions = []
....
  constructor() {

  }

```

File: src/server.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        const socket =  new WebSocketServer({ port: 8000 }); //TODO move magic numbers to env || conf
+
+        socket.on('connection', (socket) => {
+            socket.id = uuid()
+            this.connections.push(socket)
+            socket.send('Welcome to the Dash WebSocket server!');
+        });
+
+        console.log('WebSocket server is running on ws://localhost:8000');
+        setInterval(() => this.handleRawTransactions(), 100)
```
Comment (pshenmic) [inline src/server.js]: Use setTimeout that runs itself, because setInterval is not that safe. setInterval will execute your function each 100ms, but if your handler function take more time to process (that easily could be during high IO, or simply timeout), then your function will be called twice and that can lead to unexpected behaviour.

Use setTimeout to invoke your function, and then at the end execute setTimeout again. This way, your code will wait until your handler is finished, and only after that it calls it again.

Comment (pshenmic) [reply src/server.js]: Actually I think you can do it even without setInterval / setTimeout at all. Since everything is streams, we should just correctly pipe functions together.

zmq is a stream
websocket is a stream

so i think it should be possible to inject your middleware function somewhere in the between, that will decode the transaction on the fly and pass it to the socket.

you should listen for zmq event, once it arrives, decode it and submit and event in the websocket

https://nodejs.org/api/stream.html
https://nodesource.com/blog/understanding-streams-in-nodejs/

streams is a OS low level thing. your terminal bash pipes works the same (f.e `cat big_file.txt | grep 'keyword' | sed ...`)

File: src/server.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    async handleRawTransactions() {
+        if (this.transactions.length) {
+            const decodedTransaction = await decodeRawTransaction(this.transactions.pop())
+            this.messageManager.broadcastMessage(this.connections, decodedTransaction);
```
Comment (pshenmic) [inline src/server.js]: Message manager basically a 1 function file, i ain't feel like it makes a lot of sense.

Since what we do here is just sending message to the client, I think its good idea to have a Client class, we store them in the Server, same as transactions, and then just execute a method on it:
```
this.clients.forEach((client) => client.broadcastMessage(decodedTransaction))
```

File: src/server.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            socket.send('Welcome to the Dash WebSocket server!');
+        });
+
+        console.log('WebSocket server is running on ws://localhost:8000');
+        setInterval(() => this.handleRawTransactions(), 100)
+    }
+
+    async handleRawTransactions() {
+        if (this.transactions.length) {
+            const decodedTransaction = await decodeRawTransaction(this.transactions.pop())
```
Comment (pshenmic) [inline src/server.js]: We can decode a transaction locally with an dashcore-lib, it will save us reaaaly a lot of time. Making requests is expensive

Comment (pshenmic) [reply src/server.js]: Making async calls are also tricky with the streams, because streams are always synchronous

**Review by pshenmic — COMMENTED** [review]
would be good to have a good PR description that will show what exactly was changed from the developer perspective, like added ESModules, added server, zeromq, ws and all that what happened during this PR. It is the place where people will come to see what was changed recently and between the versions

also would be really cool see LICENSE (get MIT), README.md, CHANGELOG.md, .gitignore files in the root of the repo

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+# Stores VSCode versions used for testing VSCode extensions
+.vscode-test
+
+# yarn v2
+.yarn/cache
+.yarn/unplugged
+.yarn/build-state.yml
+.yarn/install-state.gz
+.pnp.*
```
Comment (pshenmic) [inline .gitignore:131]: Worth adding `.idea` for the Webstorm users, so theirs local configuration file will not checked via git

File: src/server.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    const socket = new WebSocketServer({ port: process.env.PORT })
+
+    socket.on('connection', (socket) => {
+      socket.id = uuid()
+      const client = new Client(socket)
+      this.#clients.push(client)
+      socket.send('You are successfully connected to the Dash WebSocket server!')
+    })
+
+    console.log(`WebSocket server is running on ${process.env.SERVER_URL}`)
```
Comment (pshenmic) [inline src/server.js]: I don't think you even need `SERVER_URL` env, because it will almost always differs from your external ip. Actually, I think your bound interface ip could be available as a callback arguments in the  .open() event. Not sure though about WebSocketServer, but fastify for sure returns that.

File: src/server.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  async start () {
+    const socket = new WebSocketServer({ port: process.env.PORT })
+
+    socket.on('connection', (socket) => {
+      socket.id = uuid()
+      const client = new Client(socket)
+      this.#clients.push(client)
+      socket.send('You are successfully connected to the Dash WebSocket server!')
```
Comment (pshenmic) [inline src/server.js]: Use your client to push messages

File: src/client.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      to: tx.outputs.map(output => {
+        return {
+          address: output.script?.toAddress(Dashcore.Networks.testnet).toString(),
+          valueSat: output.satoshis
+        }
+      })
+    }
+  }
+
+  broadcastMessage (message) {
```
Comment (pshenmic) [inline src/client.js]: I don't think it is a broadcast, because it is linked with the specific client.
I would name it `sendMessage`

File: src/client.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    if (this.socket.readyState === 1) {
+      this.socket.send(JSON.stringify(this.#modifiedTransaction(message)))
```
Comment (pshenmic) [inline src/client.js]: You should not map here I believe, because the function is intended to just send message to the client, it should just take the input and send it over the socket. The mapping should occur on the upper level

File: src/client.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import Dashcore from '@dashevo/dashcore-lib'
+
+export default class Client {
+  constructor (socket) {
+    this.socket = socket
+  }
+
+  #modifiedTransaction (tx) {
```
Comment (pshenmic) [inline src/client.js]: I think that this function is not really related to the Client, I would put it in the utils or better in some other class method function.

Actually, I think it could be even better if we send message to the client as is and do not modify it. Then developer can easily rely on currently existing dashcore-lib documentation (since its already typed) to work with your data.

File: src/client.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    return {
+      txId: tx.hash,
+      version: tx.version,
+      type: tx.type,
+      size: tx.toBuffer().length,
+      locktime: tx.nLockTime,
+      from: tx.inputs.map(input => input.script?.toAddress(Dashcore.Networks.testnet).toString()),
```
Comment (pshenmic) [inline src/client.js]: We should define network at the start up / init phase. Create an `NETWORK` environment, then you can use proper address for different networks

File: src/server.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+import Client from './client.js'
+import Dashcore from '@dashevo/dashcore-lib'
+
+export default class Server {
+  #clients = []
+
+  constructor () {
+    this.zmq = new ZmqConnection()
+    this.zmq.on('rawtx', (data) => {
+      this.handleRawTransactions(data)
```
Comment (pshenmic) [inline src/server.js]: `handleRawTx`

Comment (pshenmic) [reply src/server.js]: or `handleRawTransaction`

File: src/server.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      socket.id = uuid()
+      const client = new Client(socket)
+      this.#clients.push(client)
+      socket.send('You are successfully connected to the Dash WebSocket server!')
+    })
+
+    console.log(`WebSocket server is running on ${process.env.SERVER_URL}`)
+  }
+
+  handleRawTransactions (tx) {
```
Comment (pshenmic) [inline src/server.js]: `tx` -> `hex`

File: src/server.js (hunk 13 above)
Comment (pshenmic) [reply src/server.js]: also, I believe that you can use function reference instead of arrow function.

`this.zmq.on('rawtx', this.handleRawTx)`

File: src/server.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    this.zmq.on('rawtx', (data) => {
+      this.handleRawTransactions(data)
+    })
+  }
+
+  async start () {
+    const socket = new WebSocketServer({ port: process.env.PORT })
+
+    socket.on('connection', (socket) => {
+      socket.id = uuid()
```
Comment (pshenmic) [inline src/server.js]: I feel like this uuid more related to Client, rather than socket, since it is user-generated. It is also a best practice to not mutate arguments and objects you do not own (that field could be used internally by the library)

File: package.json [hunk 16]
```diff
@@ -7,8 +7,17 @@
     "start": "node index.js",
     "test": "echo \"Error: no test specified\" && exit 1"
   },
+  "type": "module",
   "private": true,
   "dependencies": {
-    "ws": "^8.18.0"
+    "@dashevo/dashcore-lib": "^0.21.3",
+    "bs58check": "^4.0.0",
```
Comment (pshenmic) [inline package.json]: This module is not used right now. And I think you can do address validation through the dashcore-lib actually. 

this is how we do it

image

File: index.js [hunk 17]
```diff
@@ -0,0 +1,3 @@
+import Server from './src/server.js'
+
+new Server().start().catch(e => console.log(e))
```
Comment (pshenmic) [inline index.js]: const server = new Server()

server.start().catch(console.error)

Comment (pshenmic) [reply index.js]: or `new Server().start().catch(console.error)`

but I believe you will add more functions to your Server, like listen(), if you start making tests, so I would keep the previous one

**Review by pshenmic — COMMENTED** [review]
More or less its getting to somewhat solid code architecture

File: src/client.js (hunk 11 above)
Comment (maxmanukian) [reply src/client.js]: I have seen websocket services structured like that, so I assumed we should do something similar. However, you’re right developers who will  use it can modify it themselves.

File: src/server.js (hunk 13 above)
Comment (maxmanukian) [reply src/server.js]: yes  it’s working, but in that case  we need to bind the class , so we can use the arrow function syntax and avoid additionally setting this.handleRaw = this.handleRaw.bind(this)

File: src/server.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    const socket = new WebSocketServer({ port: process.env.PORT })
+
+    socket.on('connection', (socket) => {
+      const client = new Client(socket)
+      this.#clients.push(client)
+      client.sendMessage('You are successfully connected to the Dash WebSocket server!')
+    })
+
+    const addressInfo = socket.address()
+    const host = addressInfo.address === '::' ? 'localhost' : addressInfo.address
```
Comment (pshenmic) [inline src/server.js]: I think this is not need, it is showing what address what exactly ip interface it was bound on.

Actually. that basically depends on what your port your application was bound on. I think you need an environment variable to set the actual IP address that you want to bind on, in case you wanna change it.
It should show correctly in this case.

Such variable will be also necessary if you want to bind on 0.0.0.0 and share the access to your service

**Review by pshenmic — COMMENTED** [review]
Would be good ot see a readme explaining how to install, run, and test your code, but the rest I think fine

File: src/server.js (hunk 18 above)
Comment (pshenmic) [reply src/server.js]: This is also important not to change output of this address so it is more clear for devops that do debug their deployments
